src/lib/utils/meta_tables/meta_system_utilization_table.cpp

```diff
@@ -0,0 +1,270 @@
+}
+
+/*
+  * Returns the time in ns since epoch.
+*/
+uint64_t MetaSystemUtilizationTable::_get_total_time() {
+  auto time = std::chrono::steady_clock::now().time_since_epoch();
+  return std::chrono::duration_cast<std::chrono::nanoseconds>(time).count();
+}
+
+/*
+ * Returns the time in ns that ALL processes have spent on the CPU
+ * since an arbitrary point in the past.
+*/
+uint64_t MetaSystemUtilizationTable::_get_system_cpu_time() {
+#ifdef __linux__
+  std::ifstream stat_file;
+  stat_file.open("/proc/stat", std::ifstream::in);
+  DebugAssert(stat_file.is_open(), "Failed to open /proc/stat");
+
+  std::string cpu_line;
+  std::getline(stat_file, cpu_line);
+  stat_file.close();
+
+  const auto cpu_ticks = _parse_value_string(cpu_line);
+
+  const auto user_ticks = cpu_ticks.at(0);
+  const auto user_nice_ticks = cpu_ticks.at(1);
+  const auto kernel_ticks = cpu_ticks.at(2);
+
+  const auto active_ticks = user_ticks + user_nice_ticks + kernel_ticks;
+  size_t cpu_count;
+  if (numa_available() != -1) {
+    cpu_count = numa_num_task_cpus();
+  } else {
+    cpu_count = _get_cpu_count();
+  }
+
+  // The amount of time in /proc/stat is measured in units of clock ticks.
+  // sysconf(_SC_CLK_TCK) can be used to convert it to ns.
+  const auto active_ns = (active_ticks * std::nano::den) / (sysconf(_SC_CLK_TCK) * cpu_count);
+
+  return active_ns;
+#endif
+
+#ifdef __APPLE__
+  host_cpu_load_info_data_t cpu_info;
+  mach_msg_type_number_t count = HOST_CPU_LOAD_INFO_COUNT;
+  [[maybe_unused]] const auto ret =
+      host_statistics(mach_host_self(), HOST_CPU_LOAD_INFO, reinterpret_cast<host_info_t>(&cpu_info), &count);
+  DebugAssert(ret == KERN_SUCCESS, "Failed to get host_statistics");
+
+  const auto active_ticks =
+      cpu_info.cpu_ticks[CPU_STATE_SYSTEM] + cpu_info.cpu_ticks[CPU_STATE_USER] + cpu_info.cpu_ticks[CPU_STATE_NICE];
+
+  // The amount of time from HOST_CPU_LOAD_INFO is measured in units of clock ticks.
+  // sysconf(_SC_CLK_TCK) can be used to convert it to ns.
+  const auto active_ns = active_ticks * std::nano::den / (sysconf(_SC_CLK_TCK) * _get_cpu_count());
+
+  return active_ns;
+#endif
+
+  Fail("Method not implemented for this platform");
+}
+
+/*
+ * Returns the time in ns that THIS process has spent on the CPU
+ * since an arbitrary point in the past.
+*/
+uint64_t MetaSystemUtilizationTable::_get_process_cpu_time() {
+  // CLOCK_PROCESS_CPUTIME_ID:
+  // Per-process CPU-time clock (measures CPU time consumed by all threads in the process).
+#ifdef __linux__
+  struct timespec time_spec {};
+  [[maybe_unused]] const auto ret = clock_gettime(CLOCK_PROCESS_CPUTIME_ID, &time_spec);
+  DebugAssert(ret == 0, "Failed in clock_gettime");
+
+  size_t cpu_count;
+  if (numa_available() != -1) {
+    cpu_count = numa_num_task_cpus();
+  } else {
+    cpu_count = _get_cpu_count();
+  }
+
+  const auto active_ns = (time_spec.tv_sec * std::nano::den + time_spec.tv_nsec) / cpu_count;

```

Comment:
Why do we divide by cpu_count here?

Reply:
That was a handy way to get the time spent on an average CPU, but you are right, the user of the meta tables should probably do this calculation and we should return the raw values.

Reply:
In order to divide it by the number of CPUs, it would have to be an aggregate value, wouldn't it? If I am not mistaken, this is simply RDTSC, i.e., the time spent since a random point in time. How does that relate to the number of CPUs?

Reply:
true. this is a leftover from the previous implementation where we measured multiple points in time.